File: web/src/engine/rules.js
// Rule sets and the player edge they imply. Mirrors src/bjtoolkit/rules.py.

import {
  BASELINE_8D,
  HITS_SOFT_17,
  DOUBLE_9_TO_11,
  DOUBLE_10_11,
  NO_DAS,
  RESPLIT_ACES,
  SPLIT_TO_2_HANDS,
  SPLIT_TO_3_HANDS,
  LATE_SURRENDER_VS_10,
  LATE_SURRENDER_VS_9_10,
  LATE_SURRENDER_ALL,
  EARLY_SURRENDER_VS_10,
  EARLY_SURRENDER_ALL,
  BJ_PAYS_6_5,
  BJ_PAYS_7_5,
  BJ_PAYS_1_1,
  ENHC_ORIGINAL_BET_ONLY,
  ENHC_ALL_BETS_LOST,
  DECK_EFFECT,
} from "./constants.js";

export const Surrender = {
  NONE: "none",
  LATE_VS_10: "late_vs_10",
  LATE_VS_9_10: "late_vs_9_10",
  LATE_ALL: "late_all",
  EARLY_VS_10: "early_vs_10",
  EARLY_ALL: "early_all",
};

export const HoleCard = {
  PEEK: "peek",
  ENHC_ORIGINAL_ONLY: "enhc_original_only",
  ENHC_ALL_BETS: "enhc_all_bets",
};

export const DoubleRule = {
  ANY_TWO: "any_two",
  NINE_TO_ELEVEN: "9_to_11",
  TEN_ELEVEN: "10_11",
};

export const isEarlySurrender = (s) =>
  s === Surrender.EARLY_VS_10 || s === Surrender.EARLY_ALL;

export const DEFAULTS = {
  name: "custom",
  decks: 6,
  dealerHitsSoft17: false,
  doubleRule: DoubleRule.ANY_TWO,
  doubleAfterSplit: true,
  resplitAces: false,
  maxSplitHands: 4,
  surrender: Surrender.NONE,
  holeCard: HoleCard.PEEK,
  blackjackPayout: 1.5,
  penetrationDecksDealt: 4.5,
  playersAtTable: 1,
  tableMin: 200,
  tableMax: 5000,
  currency: "CZK",
};

export function makeRules(overrides = {}) {
  const r = { ...DEFAULTS, ...overrides };
  if (r.penetrationDecksDealt >= r.decks) {
    r.penetrationDecksDealt = Math.max(0.5, r.decks - 0.5);
  }
  if (r.tableMax <= r.tableMin) r.tableMax = r.tableMin * 2;
  return r;
}

export const penetrationFraction = (r) => r.penetrationDecksDealt / r.decks;
export const maxTableSpread = (r) => r.tableMax / r.tableMin;

function deckEffect(decks) {
  const known = Object.keys(DECK_EFFECT).map(Number).sort((a, b) => a - b);
  if (DECK_EFFECT[decks]) return DECK_EFFECT[decks];
  const lo = Math.max(...known.filter((d) => d < decks));
  const hi = Math.min(...known.filter((d) => d > decks));
  const t = (decks - lo) / (hi - lo);
  const a = DECK_EFFECT[lo];
  const b = DECK_EFFECT[hi];
  return {
    value: a.value + t * (b.value - a.value),
    low: a.low + t * (b.low - a.low),
    high: a.high + t * (b.high - a.high),
    source: `interpolated between ${lo} and ${hi} decks`,
  };
}

export function edgeComponents(r) {
  const parts = [["baseline (8D, S17, DAS, 3:2, peek)", BASELINE_8D]];

  if (r.decks !== 8) parts.push([`${r.decks} decks instead of 8`, deckEffect(r.decks)]);
  if (r.dealerHitsSoft17) parts.push(["dealer hits soft 17", HITS_SOFT_17]);

  if (r.doubleRule === DoubleRule.NINE_TO_ELEVEN)
    parts.push(["double 9-11 only", DOUBLE_9_TO_11]);
  else if (r.doubleRule === DoubleRule.TEN_ELEVEN)
    parts.push(["double 10-11 only", DOUBLE_10_11]);

  if (!r.doubleAfterSplit) parts.push(["no double after split", NO_DAS]);
  if (r.resplitAces) parts.push(["resplit aces", RESPLIT_ACES]);
  if (r.maxSplitHands === 2) parts.push(["split to 2 hands only", SPLIT_TO_2_HANDS]);
  else if (r.maxSplitHands === 3) parts.push(["split to 3 hands only", SPLIT_TO_3_HANDS]);

  const surrenderEffect = {
    [Surrender.NONE]: null,
    [Surrender.LATE_VS_10]: ["late surrender vs 10", LATE_SURRENDER_VS_10],
    [Surrender.LATE_VS_9_10]: ["late surrender vs 9-10", LATE_SURRENDER_VS_9_10],
    [Surrender.LATE_ALL]: ["late surrender", LATE_SURRENDER_ALL],
    [Surrender.EARLY_VS_10]: ["early surrender vs 10", EARLY_SURRENDER_VS_10],
    [Surrender.EARLY_ALL]: ["early surrender", EARLY_SURRENDER_ALL],
  }[r.surrender];
  if (surrenderEffect) parts.push(surrenderEffect);

  if (r.holeCard === HoleCard.ENHC_ORIGINAL_ONLY)
    parts.push(["no hole card, original bet only", ENHC_ORIGINAL_BET_ONLY]);
  else if (r.holeCard === HoleCard.ENHC_ALL_BETS)
    parts.push(["no hole card, all bets lost", ENHC_ALL_BETS_LOST]);

  const payout = Math.round(r.blackjackPayout * 10000) / 10000;
  if (payout !== 1.5) {
    const known = { 1.2: BJ_PAYS_6_5, 1.4: BJ_PAYS_7_5, 1.0: BJ_PAYS_1_1 };
    if (known[payout]) parts.push([`blackjack pays ${payout}:1`, known[payout]]);
    else {
      const scaled = (BJ_PAYS_6_5.value * (1.5 - payout)) / 0.3;
      parts.push([`blackjack pays ${payout}:1`, {
        value: scaled, low: scaled * 1.1, high: scaled * 0.9,
        source: "scaled linearly from the 6:5 figure",
      }]);
    }
  }
  return parts;
}

export const baseEdge = (r) =>
  edgeComponents(r).reduce((acc, [, e]) => acc + e.value, 0);

export function baseEdgeRange(r) {
  const parts = edgeComponents(r);
  return [
    parts.reduce((a, [, e]) => a + e.low, 0),
    parts.reduce((a, [, e]) => a + e.high, 0),
  ];
}

export const PRESETS = {
  ambassador: makeRules({
    name: "Casino Ambassador (Prague)", decks: 6, dealerHitsSoft17: false,
    surrender: Surrender.LATE_VS_9_10, holeCard: HoleCard.ENHC_ALL_BETS,
    penetrationDecksDealt: 4.5, tableMin: 200, tableMax: 5000, currency: "CZK",
  }),
  "vegas-strip": makeRules({
    name: "Vegas Strip 6-deck (S17, DAS, LS)", decks: 6,
    surrender: Surrender.LATE_ALL, tableMin: 25, tableMax: 2000, currency: "USD",
  }),
  "vegas-h17": makeRules({
    name: "Vegas 6-deck H17", decks: 6, dealerHitsSoft17: true,
    surrender: Surrender.LATE_ALL, tableMin: 15, tableMax: 1000, currency: "USD",
  }),
  "vegas-8d": makeRules({
    name: "Vegas 8-deck H17, no surrender", decks: 8, dealerHitsSoft17: true,
    penetrationDecksDealt: 6, tableMin: 10, tableMax: 1000, currency: "USD",
  }),
  "single-deck": makeRules({
    name: "Single deck S17, double 10-11 only", decks: 1,
    doubleRule: DoubleRule.TEN_ELEVEN, doubleAfterSplit: false, maxSplitHands: 2,
    penetrationDecksDealt: 0.65, tableMin: 25, tableMax: 500, currency: "USD",
  }),
  "double-deck": makeRules({
    name: "Double deck H17, DAS", decks: 2, dealerHitsSoft17: true,
    penetrationDecksDealt: 1.5, tableMin: 25, tableMax: 1000, currency: "USD",
  }),
  "six-five": makeRules({
    name: "6:5 tourist table -- do not play this", decks: 6,
    dealerHitsSoft17: true, blackjackPayout: 1.2, penetrationDecksDealt: 4,
    tableMin: 10, tableMax: 500, currency: "USD",
  }),
  "uk-enhc": makeRules({
    name: "UK/European 6-deck ENHC, S17", decks: 6,
    holeCard: HoleCard.ENHC_ALL_BETS, penetrationDecksDealt: 4.5,
    tableMin: 10, tableMax: 500, currency: "GBP",
  }),
};

File: src/bjtoolkit/rules.py
"""Rule sets for any blackjack game, and the player edge they imply.

The design goal is that a game anywhere in the world can be described here:
Czech ENHC, Vegas Strip, downtown single deck, a 6:5 tourist trap. `base_edge`
is an additive model over `constants.py`, so every toggle moves the number and
nothing is baked in for one casino.
"""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field, model_validator

from . import constants as C
from .constants import Estimate


class Surrender(str, Enum):
    NONE = "none"
    LATE_VS_10 = "late_vs_10"
    LATE_VS_9_10 = "late_vs_9_10"
    LATE_ALL = "late_all"
    EARLY_VS_10 = "early_vs_10"
    EARLY_ALL = "early_all"

    @property
    def is_early(self) -> bool:
        return self in (Surrender.EARLY_VS_10, Surrender.EARLY_ALL)


class HoleCard(str, Enum):
    """How the dealer's second card is handled."""

    PEEK = "peek"
    """US style: dealer takes a hole card and checks it for blackjack."""

    ENHC_ORIGINAL_ONLY = "enhc_original_only"
    """No hole card, but only your original bet is lost to a dealer blackjack."""

    ENHC_ALL_BETS = "enhc_all_bets"
    """No hole card, and doubled/split bets are lost too. Needs strategy changes."""

    @property
    def is_enhc(self) -> bool:
        return self is not HoleCard.PEEK


class DoubleRule(str, Enum):
    ANY_TWO = "any_two"
    NINE_TO_ELEVEN = "9_to_11"
    TEN_ELEVEN = "10_11"


class RuleSet(BaseModel):
    """A complete description of one blackjack game."""

    name: str = "custom"

    decks: int = Field(6, ge=1, le=8)
    dealer_hits_soft_17: bool = False
    double_rule: DoubleRule = DoubleRule.ANY_TWO
    double_after_split: bool = True
    resplit_aces: bool = False
    max_split_hands: int = Field(4, ge=2, le=4)
    surrender: Surrender = Surrender.NONE
    hole_card: HoleCard = HoleCard.PEEK
    blackjack_payout: float = Field(1.5, gt=0)

    penetration_decks_dealt: float = Field(4.5, gt=0)
    """Decks dealt before the shuffle. The single most leveraged rule there is."""

    players_at_table: int = Field(1, ge=1, le=7)
    """Including you. Sets how many cards a round burns, so how many betting
    decisions you get per shoe."""

    table_min: float = 200.0
    table_max: float = 5000.0
    currency: str = "CZK"

    @model_validator(mode="after")
    def _check(self) -> RuleSet:
        if self.penetration_decks_dealt >= self.decks:
            raise ValueError(
                f"penetration_decks_dealt ({self.penetration_decks_dealt}) must be "
                f"less than decks ({self.decks}) -- you cannot deal the whole shoe"
            )
        if self.table_max <= self.table_min:
            raise ValueError("table_max must exceed table_min")
        return self

    @property
    def penetration_fraction(self) -> float:
        return self.penetration_decks_dealt / self.decks

    @property
    def max_table_spread(self) -> float:
        return self.table_max / self.table_min


def _deck_effect(decks: int) -> Estimate:
    """Interpolate the deck-count effect for counts not in the published table."""
    if decks in C.DECK_EFFECT:
        return C.DECK_EFFECT[decks]
    known = sorted(C.DECK_EFFECT)
    lo = max(d for d in known if d < decks)
    hi = min(d for d in known if d > decks)
    t = (decks - lo) / (hi - lo)
    a, b = C.DECK_EFFECT[lo], C.DECK_EFFECT[hi]
    return Estimate(
        a.value + t * (b.value - a.value),
        a.low + t * (b.low - a.low),
        a.high + t * (b.high - a.high),
        f"interpolated between {lo} and {hi} decks",
    )


def edge_components(rules: RuleSet) -> list[tuple[str, Estimate]]:
    """The additive breakdown of the player's off-the-top edge.

    Returned as an ordered list so the CLI can show its work -- the point of the
    tool is that you can see which rule is costing you what.
    """
    parts: list[tuple[str, Estimate]] = [
        ("baseline (8D, S17, DAS, 3:2, peek)", C.BASELINE_8D),
    ]

    if rules.decks != 8:
        parts.append((f"{rules.decks} decks instead of 8", _deck_effect(rules.decks)))

    if rules.dealer_hits_soft_17:
        parts.append(("dealer hits soft 17", C.HITS_SOFT_17))

    if rules.double_rule is DoubleRule.NINE_TO_ELEVEN:
        parts.append(("double 9-11 only", C.DOUBLE_9_TO_11))
    elif rules.double_rule is DoubleRule.TEN_ELEVEN:
        parts.append(("double 10-11 only", C.DOUBLE_10_11))

    if not rules.double_after_split:
        parts.append(("no double after split", C.NO_DAS))

    if rules.resplit_aces:
        parts.append(("resplit aces", C.RESPLIT_ACES))

    if rules.max_split_hands == 2:
        parts.append(("split to 2 hands only", C.SPLIT_TO_2_HANDS))
    elif rules.max_split_hands == 3:
        parts.append(("split to 3 hands only", C.SPLIT_TO_3_HANDS))

    surrender_effect = {
        Surrender.NONE: None,
        Surrender.LATE_VS_10: ("late surrender vs 10", C.LATE_SURRENDER_VS_10),
        Surrender.LATE_VS_9_10: ("late surrender vs 9-10", C.LATE_SURRENDER_VS_9_10),
        Surrender.LATE_ALL: ("late surrender", C.LATE_SURRENDER_ALL),
        Surrender.EARLY_VS_10: ("early surrender vs 10", C.EARLY_SURRENDER_VS_10),
        Surrender.EARLY_ALL: ("early surrender", C.EARLY_SURRENDER_ALL),
    }[rules.surrender]
    if surrender_effect is not None:
        parts.append(surrender_effect)

    if rules.hole_card is HoleCard.ENHC_ORIGINAL_ONLY:
        parts.append(("no hole card, original bet only", C.ENHC_ORIGINAL_BET_ONLY))
    elif rules.hole_card is HoleCard.ENHC_ALL_BETS:
        parts.append(("no hole card, all bets lost", C.ENHC_ALL_BETS_LOST))

    payout = round(rules.blackjack_payout, 4)
    if payout != 1.5:
        known = {1.2: C.BJ_PAYS_6_5, 1.4: C.BJ_PAYS_7_5, 1.0: C.BJ_PAYS_1_1}
        if payout in known:
            parts.append((f"blackjack pays {payout}:1", known[payout]))
        else:
            # Linear in the payout shortfall; a natural occurs ~4.75% of hands.
            scaled = C.BJ_PAYS_6_5.value * (1.5 - payout) / 0.3
            parts.append((
                f"blackjack pays {payout}:1",
                Estimate(scaled, scaled * 1.1, scaled * 0.9,
                         "scaled linearly from the 6:5 figure"),
            ))

    return parts


def base_edge(rules: RuleSet) -> float:
    """Player edge off the top, before counting. Negative means house favoured."""
    return sum(e.value for _, e in edge_components(rules))


def base_edge_range(rules: RuleSet) -> tuple[float, float]:
    """Worst and best case edge, given the uncertainty on each component."""
    parts = edge_components(rules)
    return (sum(e.low for _, e in parts), sum(e.high for _, e in parts))


# ---------------------------------------------------------------------------
# Presets
# ---------------------------------------------------------------------------

PRESETS: dict[str, RuleSet] = {
    "ambassador": RuleSet(
        name="Casino Ambassador (Prague) -- UNVERIFIED, see docs/open-questions.md",
        decks=6,
        dealer_hits_soft_17=False,
        double_after_split=True,
        resplit_aces=False,
        max_split_hands=4,
        surrender=Surrender.LATE_VS_9_10,
        hole_card=HoleCard.ENHC_ALL_BETS,
        penetration_decks_dealt=4.5,
        table_min=200.0,
        table_max=5000.0,
        currency="CZK",
    ),
    "vegas-strip": RuleSet(
        name="Vegas Strip 6-deck (S17, DAS, LS)",
        decks=6, dealer_hits_soft_17=False, surrender=Surrender.LATE_ALL,
        penetration_decks_dealt=4.5, table_min=25.0, table_max=2000.0,
        currency="USD",
    ),
    "vegas-h17": RuleSet(
        name="Vegas 6-deck H17 (the common downtown game)",
        decks=6, dealer_hits_soft_17=True, surrender=Surrender.LATE_ALL,
        penetration_decks_dealt=4.5, table_min=15.0, table_max=1000.0,
        currency="USD",
    ),
    "vegas-8d": RuleSet(
        name="Vegas 8-deck H17, no surrender",
        decks=8, dealer_hits_soft_17=True, surrender=Surrender.NONE,
        penetration_decks_dealt=6.0, table_min=10.0, table_max=1000.0,
        currency="USD",
    ),
    "single-deck": RuleSet(
        name="Single deck S17, double 10-11 only",
        decks=1, dealer_hits_soft_17=False, double_rule=DoubleRule.TEN_ELEVEN,
        double_after_split=False, max_split_hands=2,
        penetration_decks_dealt=0.65, table_min=25.0, table_max=500.0,
        currency="USD",
    ),
    "double-deck": RuleSet(
        name="Double deck H17, DAS",
        decks=2, dealer_hits_soft_17=True, penetration_decks_dealt=1.5,
        table_min=25.0, table_max=1000.0, currency="USD",
    ),
    "six-five": RuleSet(
        name="6:5 tourist table -- do not play this",
        decks=6, dealer_hits_soft_17=True, blackjack_payout=1.2,
        penetration_decks_dealt=4.0, table_min=10.0, table_max=500.0,
        currency="USD",
    ),
    "uk-enhc": RuleSet(
        name="UK/European 6-deck ENHC, S17",
        decks=6, dealer_hits_soft_17=False, hole_card=HoleCard.ENHC_ALL_BETS,
        surrender=Surrender.NONE, penetration_decks_dealt=4.5,
        table_min=10.0, table_max=500.0, currency="GBP",
    ),
}


def get_preset(name: str) -> RuleSet:
    try:
        return PRESETS[name].model_copy(deep=True)
    except KeyError:
        raise KeyError(
            f"unknown preset {name!r}; available: {', '.join(sorted(PRESETS))}"
        ) from None
